### ISSUE TITLE
How to authenticate without login route ?

### ISSUE THREAD

**Issue by bakura10** [heartsentwined/ember-auth#10]
Hi,

I'm using EmberJS for a small subset of my ZF 2 application. Therefore, the user is authenticated through server side page, hence bypassing the mechanism introduced by this Ember module.

I'm a beginner in REST architecture, but how to deal with that ? Currently I save auth_token and user_id on the server instead of returning them in Json. The problme is that your lbirary save authToken and currentUserId as a return of signIn Ember route.

What is the best way to achieve this ?

**Comment by bakura10** [comment]
Furthermore, is there a way the token is not sent through the "Authorization" header instead as a GET parameter ?

**Comment by heartsentwined** [comment]
I don't understand the use case you described. If the user is already authenticated in a "server side page", then this happens before the ember app is loaded, right? If this is the case, then the user will be always "signed in" in the ember app, what are you trying achieve with a client-side authentication solution? Or have I read your use case wrong?

`ember-auth` does make assumptions in that you follow the REST architecture for API end points, hence it does expect `auth_token` (among others) from your replies. I would suggest that you implement the suggested API end points, since the rest of your ember app will most likely use REST-style APIs to communicate with your server anyway.

I also don't understand the "authorization" header part. `ember-auth` does send the authentication token as a GET parameter. What is this "authorization" header you are referring to?

**Comment by bakura10** [comment]
Hi,

In fact in my application the user logged through server side. But I've found another solution (that rely on session identifier).

However concerning the authorization part I think you're doing it wrong. REST is making use of HTTP at its maximum... And there is a header specificially made for authentication: the Authorization header (http://en.wikipedia.org/wiki/Basic_access_authentication).

Basically, instead of giving the token to a GET parameter, you should add an Authorization header, whose value is Basic auth_token

**Comment by heartsentwined** [comment]
The Authorization header approach is, in my opinion, not ideal. In particular:
- so we send username and password on each request, then how do we differentiate the initial sign in from subsequent authenticated accesses? (see next point)
- there are one-off actions that should be performed only on initial sign in: increment a sign-in count, log IP, set remember me period, etc
- what about sign out? There is no de-Authorization header as far as I know

The current implementation (following [devise](http://devise.plataformatec.com.br/) of [rails](http://rubyonrails.org/)) is to treat sign-in and sign-out actions as one of the REST models.

A simplified server side implemention is to declare a `session` resource, and allow `create` and `destroy` actions on it:
- `POST session`: creates a session, i.e. sign in. As with other REST resource create actions, the model parameters are passed as POST parameters.
- `DELETE session`: destroys a session, i.e. sign out. The auth token is passed along not only as a session identifier, but also as an authentication mechanism: a stranger should not be allowed to destroy others' session. A standard RESTful route would be `DELETE session/(auth token)`, but we choose to follow the authentication convention of other API calls. (read along next paragraph)

The design decision to pass an auth token on all subsequent model actions is based on existing API services practices. Most API services expect exactly the same behavior:
1. request an auth token at a dedicated API end point
2. give this auth token along with other parameters when accessing other API end points

This way your server side app exposes a standard API, your ember app is just one client side app that consumes this API, and your app can then be consumed additionally by other client side implementations. An example is to write a smartphone without ember, but using the same authentication mechanism.

You may want to read up on RESTful architecture to understand this mechanism more:
- [a general overview](http://www.infoq.com/articles/rest-introduction)
- [wikipedia](http://en.wikipedia.org/wiki/REST)
- [the Rails implementation](http://guides.rubyonrails.org/routing.html#resource-routing-the-rails-default)

**Comment by bakura10** [comment]
I'll read carefully the other links, but I can't see why your reasons does not allow to use the Authorization header (see that in my previous post - that I edited - I made the mistake too... there are both Authentication AND Authorization header, the Authorization one is for sending the token). You can still take advantage of your POST session and DELETE session, but still send the token through the Authorization header instead as a GET (or make it as an option maybe ?).

But well, I'm not a REST expert so I won't say more about it ;-).

**Comment by heartsentwined** [comment]
Ah, I got confused like you did; edited above post likewise :-)

[Amazon's solution](http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html) seems the best example use case for Authorization headers. Marked as enhancement for future implementation.

**Comment by bakura10** [comment]
Exactly. Their solution is a bit more complex but the idea is here: Authorization header ;-). I'll keep an eye on it. Anyway, good job on this library, looks cool !

**Comment by heartsentwined** [comment]
@bakura10 Half of amazon's solution has now been implemented in the `3.0` branch. You can now choose to use the Authorization header, or a custom header, to send the authentication token, instead of passing it as a param.

The HMAC encryption has not been implemented yet.

BC Breaks and upgrade guide for `3.0` branch available in the [Changelog](https://github.com/heartsentwined/ember-auth/blob/master/CHANGELOG.md).

Full relevant docs section from readme:

> ### Different token locations
> 
> `ember-auth` supports three different methods of sending the authentication
> token on API requests:
> - include it as one of the key-value pairs in the data/params hash (default)
> - use the Authorization header ([RFC 1945](http://tools.ietf.org/html/rfc1945#section-10.2))
> - send it in a custom header
> 
> #### Key-value pair in data/params hash
> 
> Send `auth_token = auth_token_value` alongside your other API request params:
> 
> ``` coffeescript
> Auth.Config.reopen
>   requestTokenLocation: 'param' # or omit this line - this is the default value
>   tokenKey: 'auth_token'
> ```
> 
> #### Authorization header
> 
> Send a `Authorization: TOKEN auth_token_value` header:
> 
> ``` coffeescript
> Auth.Config.reopen
>   requestTokenLocation: 'authHeader'
>   requestHeaderKey: 'TOKEN'
> ```
> 
> #### Custom header
> 
> Send a `X-API-TOKEN: auth_token_value` header:
> 
> ``` coffeescript
> Auth.Config.reopen
>   requestTokenLocation: 'customHeader'
>   requestHeaderKey: 'X-API-TOKEN'
> ```

**Comment by bakura10** [comment]
This is looking awesome, thanks a lot for implemeting this so quickly :). Unfortunately I could not test it because I've done something much simpler in my application, but as my next application will be 100% EmberJS I'll use this !

About naming, I think 'requestTokenLocation' may be renamed to 'requestTokenMethod', it's clearer and better represent what it does.

Here the "" at the beginning seems useless : https://github.com/heartsentwined/ember-auth/blob/master/dist/ember-auth.js#L218

**Comment by heartsentwined** [comment]
The `-location` suffix is intentional, exactly because I have "reserved" the `-method` suffix in mind for future implementation of different hashing/security mechanisms on the token.

As for that beginning `""`, well, that was a generated distribution file from coffeescript sources. But I think it is not useless: it coerces the whole thing into a string, because the `+` operator is polymorphic.
